File: frontend/src/api/projects.ts
import { api } from "@/api/client";

export type ProjectRole = "owner" | "editor" | "viewer";

export interface ProjectListItem {
  id: string;
  name: string;
  description: string;
  role: ProjectRole;
  member_count: number;
  created_at: string;
}

export interface Member {
  user_id: string;
  email: string;
  name: string;
  role: ProjectRole;
  is_owner: boolean;
}

export interface ProjectDetail {
  id: string;
  name: string;
  description: string;
  owner_id: string;
  created_at: string;
  role: ProjectRole;
  members: Member[];
}

export const Projects = {
  list:    ()  => api.get<ProjectListItem[]>("/projects"),
  get:     (id: string) => api.get<ProjectDetail>(`/projects/${id}`),
  create:  (body: { name: string; description?: string }) =>
    api.post<ProjectListItem>("/projects", body),
  update:  (id: string, body: { name?: string; description?: string }) =>
    api.patch<ProjectDetail>(`/projects/${id}`, body),
  delete:  (id: string) => api.del<void>(`/projects/${id}`),
  members: {
    list:   (id: string) => api.get<Member[]>(`/projects/${id}/members`),
    add:    (id: string, body: { email: string; role: ProjectRole }) =>
      api.post<Member>(`/projects/${id}/members`, body),
    update: (id: string, userId: string, role: ProjectRole) =>
      api.patch<Member>(`/projects/${id}/members/${userId}`, { role }),
    remove: (id: string, userId: string) =>
      api.del<void>(`/projects/${id}/members/${userId}`),
  },
};

File: backend/app/schemas/projects.py
"""Pydantic shapes for projects + memberships."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, EmailStr, Field

ProjectRole = Literal["owner", "editor", "viewer"]


class ProjectCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)


class ProjectUpdateRequest(BaseModel):
    name: str | None = Field(default=None, max_length=200)
    description: str | None = Field(default=None, max_length=2000)


class ProjectListItem(BaseModel):
    id: uuid.UUID
    name: str
    description: str
    role: ProjectRole          # the *requester's* role on this project
    member_count: int
    created_at: datetime


class MemberItem(BaseModel):
    user_id: uuid.UUID
    email: EmailStr
    name: str
    role: ProjectRole
    is_owner: bool


class ProjectDetail(BaseModel):
    id: uuid.UUID
    name: str
    description: str
    owner_id: uuid.UUID
    created_at: datetime
    role: ProjectRole          # the *requester's* role
    members: list[MemberItem]


class MemberAddRequest(BaseModel):
    email: EmailStr
    role: ProjectRole = "editor"


class MemberRoleUpdate(BaseModel):
    role: ProjectRole
